pkg/lightning/errormanager/errormanager.go

```diff
@@ -503,178 +513,284 @@ func (em *ErrorManager) ReplaceConflictKeys(
+							zap.Binary("rawValue", rawValue))
+
+						// If rawKey equals to KV pair's key and rawValue equals to KV pair's value,
+						// this latest data KV of the index KV needs to be deleted;
+						// if not, this latest data KV of the index KV was inserted by other rows,
+						// so it is unrelated to the index KV that needs to be deleted, we cannot delete it.
+
+						// An example is:
+						// (pk, uk)
+						// (1, a)
+						// (1, b)
+						// (2, a)
+
+						// (1, a) is overwritten by (2, a). We found a->1 is an overwritten index KV,
+						// and we are considering if its data KV with key "1" can be deleted.
+						// We got the latest value of key "1" which is (1, b),
+						// and encode it to get all KV pairs which is [1->b, b->1].
+						// Only if there is a->1 we dare to delete data KV with key "1".
+
+						if bytes.Equal(kvPair.Key, rawKey) && bytes.Equal(kvPair.Val, rawValue) {
+							handleKeys = append(handleKeys, rawHandle)
+							var insertRow [2][]byte
+							insertRow[0] = rawHandle
+							insertRow[1] = overwritten
+							insertRows = append(insertRows, insertRow)
+							break
+						}
+					}
+				}
+				if err := indexKvRows.Err(); err != nil {
+					return errors.Trace(err)
+				}
+				if err := indexKvRows.Close(); err != nil {
+					return errors.Trace(err)
+				}
+				if len(handleKeys) == 0 {
+					break
+				}
+				if err := fnDeleteKeys(indexGCtx, handleKeys); err != nil {
+					return errors.Trace(err)
+				}
+				if err := exec.Transact(ctx, "insert data conflict error record for conflict detection 'replace' mode",
+					func(c context.Context, txn *sql.Tx) error {
+						sb := &strings.Builder{}
+						_, err2 := common.FprintfWithIdentifiers(sb, insertIntoConflictErrorData, em.schema)
+						if err2 != nil {
+							return errors.Trace(err2)
+						}
+						var sqlArgs []any
+						for i, insertRow := range insertRows {
+							if i > 0 {
+								sb.WriteByte(',')
 							}
-							var sqlArgs []any
 							sb.WriteString(sqlValuesConflictErrorData)
 							sqlArgs = append(sqlArgs,
 								em.taskID,
 								tableName,
 								nil,
 								nil,
-								rawHandle,
-								overwritten,
+								insertRow[0],
+								insertRow[1],
 								1,
 							)
-							_, err := txn.ExecContext(c, sb.String(), sqlArgs...)
-							return errors.Trace(err)
-						}); err != nil {
-						return errors.Trace(err)
-					}
-					if err := fnDeleteKey(gCtx, rawHandle); err != nil {
+						}
+						_, err := txn.ExecContext(c, sb.String(), sqlArgs...)
 						return errors.Trace(err)
+					}); err != nil {
+					return errors.Trace(err)
+				}
+				start = lastRowID + 1
+				// If the remaining tasks cannot be processed at once, split the task
+				// into two subtasks and send one of them to the other idle worker if possible.
+				if end-start > rowLimit {
+					mid := start + (end-start)/2
+					indexTaskWg.Add(1)
+					select {
+					case indexTaskCh <- [2]int64{mid, end}:
```

Comment:
it's easy to get deadlock as worker pool is of fixed size, suppose pool size is 1
- for the first split of worker 1, we can send to the channel
- the task send loop get a task but it cannot start another routine
- then worker 1 try to split another time, we cannot send

Reply:
The code structure is copied from `RemoveAllConflictKeys` in v7.5
https://github.com/pingcap/tidb/blob/e406d5780b18a1f2aaf6230cde4b6403991e228d/br/pkg/lightning/errormanager/errormanager.go#L452-L530

Reply:
This case will not happen because:
For step 1 and 2, there is no problem.
For step 3, if worker 1 fails to send the new task to `indexTaskCh`, it will go to `default` to execute `indexTaskWg.Done()` instead of sending the task, which will not cause deadlock.